scipy/integrate/_ivp/ivp.py

```diff
@@ -0,0 +1,466 @@
+           and its Applications, 13, pp. 117-120, 1974.
+    """
+    if method not in METHODS and not issubclass(method, OdeSolver):
+        raise ValueError("`method` must be one of {} or OdeSolver instance."
+                         .format(METHODS))
+
+    t0, tf = float(t_span[0]), float(t_span[1])
+
+    if t_eval is not None:
+        t_eval = np.asarray(t_eval)
+        if t_eval.ndim != 1:
+            raise ValueError("`t_eval` must be 1-dimensional.")
+
+        if np.any(t_eval < min(t0, tf)) or np.any(t_eval > max(t0, tf)):
+            raise ValueError("Values in `t_eval` are not within `t_span`")
+
+        d = np.diff(t_eval)
+        if tf > t0 and np.any(d <= 0) or tf < t0 and np.any(d >= 0):
+            raise ValueError("Values in `t_eval` are not properly sorted.")
+        t_eval_i = 0
+        n_eval = t_eval.shape[0]
+
+    if not callable(method):
+        method = METHODS[method]
+
+    solver = method(fun, t0, y0, tf, vectorized=vectorized, **options)
+
+    if t_eval is None:
+        ts = [t0]
+        ys = [y0]
+    else:
+        ts = []
+        ys = []
+
+        if len(t_eval) > 0 and t_eval[0] == t0:
+            pass
+            # If the first t_eval is the start time, deal with it now
+            ts.append(solver.t)
+            ys.append(np.expand_dims(solver.y, axis=1))
+            t_eval_i += 1
+
+    interpolants = []
+
+    events, is_terminal, event_dir = prepare_events(events)
+
+    if events is not None:
+        g = [event(t0, y0) for event in events]
+        t_events = [[] for _ in range(len(events))]
+    else:
+        t_events = None
+
+    status = None
+    s = solver.direction
+    while status is None:
+        message = solver.step()
+
+        if solver.status == 'finished':
+            status = 0
+        elif solver.status == 'failed':
+            status = -1
+            break
+
+        t_old = solver.t_old
+        t = solver.t
+        y = solver.y
+
+        if dense_output:
+            sol = solver.dense_output()
+            interpolants.append(sol)
+        else:
+            sol = None
+
+        if events is not None:
+            g_new = [event(t, y) for event in events]
+            active_events = find_active_events(g, g_new, event_dir)
+            if active_events.size > 0:
+                if sol is None:
+                    sol = solver.dense_output()
+                root_indices, roots, terminate = handle_events(
+                    sol, events, active_events, is_terminal, t_old, t)
+
+                for e, te in zip(root_indices, roots):
+                    t_events[e].append(te)
+
+                if terminate:
+                    status = 1
+                    t = roots[-1]
+                    y = sol(t)
+            g = g_new
+
+        if t_eval is None:
+            ts.append(t)
+            ys.append(y)
+        else:
+            t_step = []
+
+            while t_eval_i < n_eval and s * (t_eval[t_eval_i] - t) < 0:
```

Comment:
In terms of algorithmic complexity it is not a bottle neck, but this item-by-item appends are unnecessary. I changed it as to just increment the final index and then save the slice of the array. I don't like using vectorized operation (like np.where(...)) because it would ruin the algorithmic complexity.